pkg/agent/common/sigstore/sigstore.go

```diff
@@ -0,0 +1,478 @@
+			getFulcioIntermediates:  fulcioroots.GetIntermediates,
+			getRekorPublicKeys:      cosign.GetRekorPubs,
+			getCTLogPublicKeys:      cosign.GetCTLogPubs,
+		},
+	}
+
+	if verifier.config.Logger == nil {
+		verifier.config.Logger = hclog.Default()
+	}
+
+	if verifier.config.RekorURL == "" {
+		verifier.config.RekorURL = publicRekorURL
+	}
+
+	verifier.allowedIdentities = processAllowedIdentities(config.AllowedIdentities)
+
+	for registry, creds := range config.RegistryCredentials {
+		if creds != nil && creds.Username != "" && creds.Password != "" {
+			authOption := remote.WithAuth(&authn.Basic{
+				Username: creds.Username,
+				Password: creds.Password,
+			})
+			verifier.authOptions[registry] = authOption
+		}
+	}
+
+	return verifier
+}
+
+// Init prepares the verifier by retrieving the Fulcio certificates and Rekor and CT public keys.
+func (v *ImageVerifier) Init(ctx context.Context) error {
+	var err error
+	v.fulcioRoots, err = v.hooks.getFulcioRoots()
+	if err != nil {
+		return fmt.Errorf("failed to get fulcio root certificates: %w", err)
+	}
+
+	v.fulcioIntermediates, err = v.hooks.getFulcioIntermediates()
+	if err != nil {
+		return fmt.Errorf("failed to get fulcio intermediate certificates: %w", err)
+	}
+
+	if !v.config.IgnoreTlog {
+		v.rekorPublicKeys, err = v.hooks.getRekorPublicKeys(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get rekor public keys: %w", err)
+		}
+		v.rekorClient, err = v.hooks.getRekorClient(v.config.RekorURL, client.WithLogger(v.config.Logger))
+		if err != nil {
+			return fmt.Errorf("failed to get rekor client: %w", err)
+		}
+	}
+
+	if !v.config.IgnoreSCT {
+		v.ctLogPublicKeys, err = v.hooks.getCTLogPublicKeys(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get CT log public keys: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// Verify validates image's signatures, attestations, and transparency logs using Cosign and Rekor.
+// The imageID parameter is expected to be in the format "repository@sha256:digest".
+// It returns selectors based on the image signature and rekor bundle details.
+// Cosign ensures the image's signature issuer and subject match the configured allowed identities.
+// If the image is in the skip list, it bypasses verification and returns an empty list of selectors.
+// Uses a cache to avoid redundant verifications.
+// An error is returned if the verification of the images signatures or attestations fails.
+func (v *ImageVerifier) Verify(ctx context.Context, imageID string) ([]string, error) {
+	v.config.Logger.Debug("Verifying image with sigstore", "image_id", imageID)
+
+	// Check if the image is in the list of excluded images to determine if verification should be bypassed.
+	if v.isSkippedImage(imageID) {
+		// Return an empty list, indicating no verification was performed.
+		return []string{}, nil
+	}
+
+	// Check the cache for previously verified selectors.
+	if cachedSelectors, ok := v.verificationCache.Load(imageID); ok {
+		if cachedSelectors != nil {
```

Comment:
No, it's not possible to have an empty list of selectors. If a Sigstore verification is successful, it will always include at least the selector `image-signature:verified`.